File: 백준/Silver/14501. 퇴사/퇴사.py
import sys
input = sys.stdin.readline


def solve(day, money):

    global max_money
    
    # 근무 일수를 넘어가면 최대이익 갱신
    if day >= n:
        max_money = max(max_money, money)
        return
    
    # 상담 O: 최대 근무 일수를 넘어가지 않는 선에서
    if day + T[day] <= n:
        solve(day + T[day], money + P[day])

    # 상담 X
    solve(day + 1, money)


n = int(input())
T = []
P = []
for _ in range(n):
    d, m = map(int, input().split())
    T.append(d)
    P.append(m)

max_money = 0
solve(0, 0)

print(max_money)

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.*;

public class Main {

    static int n;
    static int[] T;
    static int[] P;
    static int maxMoney;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        T = new int[n];
        P = new int[n];

        for (int i = 0; i < n; i++) {
            String[] line = br.readLine().split(" ");
            T[i] = Integer.parseInt(line[0]);
            P[i] = Integer.parseInt(line[1]);
        }

        maxMoney = 0;
        solve(0, 0);
        System.out.println(maxMoney);
    }

    static void solve(int day, int money) {

        // 최대 일수 도달하면 maxMoney 갱신
        if (day >= n) {
            maxMoney = Math.max(money, maxMoney);
            return;
        }

        // 상담 0
        if (day + T[day] <= n) {
            solve(day + T[day], money + P[day]);
        }

        // 상담 X
        solve(day + 1, money);
    }
}